File: cpp/questions/topKFrequent.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> hm;
        for (const auto& n: nums) {
            hm[n]++;
        }
        struct cmp {
            bool operator()(const std::pair<int, int>& a, const std::pair<int ,int>& b) const {
                return a.first > b.first;
            }
        };
        std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, cmp> pq;
        for (auto it=hm.begin(); it!=hm.end(); ++it) {
            auto [n, c] = *it;
            if (pq.size() < k) {
                pq.emplace(c, n);
            } else if (c > pq.top().first) {
                pq.pop();
                pq.emplace(c, n);
            }
        }
        std::vector<int> ans;
        ans.reserve(k);
        while (!pq.empty()) {
            ans.push_back(pq.top().second);
            pq.pop();
        }
        return ans;
    }
};

File: python/questions/topKFrequent.py
from typing import List
from collections import defaultdict
import heapq

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        dt = defaultdict(int)
        for n in nums:
            dt[n] += 1
        hq = []
        for n, c in dt.items():
            if len(hq) < k:
                heapq.heappush(hq, (c, n))
            elif c > hq[0][0]:
                heapq.heappop(hq)
                heapq.heappush(hq, (c, n))
        return [n for _, n in hq]
            
